KeyframeMsg[] keyframes
int32 num_keyframes